### PR TITLE
Fix: handle all cases for consistent lookup unique on single transaction mode

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description

This PR enhances the consistent_lookup implementation around dangling rows when the transaction mode is set to `SINGLE`.
Now, it can handle more cases and allow the transaction to complete without throwing multi-db error, until an actual dml happens 2 or more shards.

## Related Issue(s)

- Closes https://github.com/vitessio/vitess/issues/6518

## Checklist

-   [X] Tests were added or are not required
-   [X] Documentation was added or is not required

### CODE REVIEW

File: go/test/endtoend/vtgate/transaction/single/main_test.go [hunk 1]
```diff
@@ -0,0 +1,225 @@
+/*
+Copyright 2019 The Vitess Authors.
```
Comment (frouioui) [inline go/test/endtoend/vtgate/transaction/single/main_test.go:2]: nit: 2022

File: go/test/endtoend/vtgate/transaction/single/main_test.go [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+			Host: clusterInstance.Hostname,
+			Port: clusterInstance.VtgateMySQLPort,
+		}
+		return m.Run()
+	}()
+	os.Exit(exitCode)
+}
+
+func TestSingleOneWay(t *testing.T) {
+	conn, err := mysql.Connect(context.Background(), &vtParams)
```
Comment (frouioui) [inline go/test/endtoend/vtgate/transaction/single/main_test.go:89]: I think this test and the one below should be tested against both Vitess and mysql, we can use the `utils` package for that.

File: go/vt/vtgate/safe_session.go [hunk 3]
```diff
@@ -294,10 +299,21 @@ func (session *SafeSession) Find(keyspace, shard string, tabletType topodatapb.T
 	}
 	for _, shardSession := range sessions {
 		if keyspace == shardSession.Target.Keyspace && tabletType == shardSession.Target.TabletType && shard == shardSession.Target.Shard {
-			return shardSession.TransactionId, shardSession.ReservedId, shardSession.TabletAlias
+			if getTxMode() == vtgatepb.TransactionMode_SINGLE && shardSession.VindexOnly && !session.queryFromVindex {
+				count := actualNoOfShardSession(session.ShardSessions)
+				// If the count of shard session which are non vindex only is greater than 0, then it is a
```
Comment (frouioui) [inline go/vt/vtgate/safe_session.go:304]: nit: `then it is an error`

File: go/vt/vtgate/vcursor_impl.go [hunk 4]
```diff
@@ -556,9 +556,18 @@ func (vc *vcursorImpl) ExecuteKeyspaceID(keyspace string, ksid []byte, query str
 		Sql:           query,
 		BindVariables: bindVars,
 	}}
+	if !rollbackOnError {
+		vc.safeSession.queryFromVindex = true
+		defer func() {
+			vc.safeSession.queryFromVindex = false
+		}()
+	}
```
Comment (frouioui) [inline go/vt/vtgate/vcursor_impl.go:564]: I don't understand this change. Maybe we should add a comment that explains why we do this

File: go/vt/vtgate/vcursor_impl.go [hunk 5]
```diff
@@ -556,9 +556,18 @@ func (vc *vcursorImpl) ExecuteKeyspaceID(keyspace string, ksid []byte, query str
...
 	qr, errs := vc.ExecuteMultiShard(rss, queries, rollbackOnError, autocommit)
-
-	return qr, vterrors.Aggregate(errs)
+	err = vterrors.Aggregate(errs)
+	if err != nil {
+		return nil, vterrors.Wrap(err, "I know I know")
```
Comment (frouioui) [inline go/vt/vtgate/vcursor_impl.go:568]: We should change the error message here

Comment (frouioui) [reply go/vt/vtgate/vcursor_impl.go:568]: Or directly return the error that `vterrors.Aggregate` returns

**Review by frouioui — APPROVED** [review]

File: go/vt/vtgate/vcursor_impl.go (hunk 5 above)
Comment (harshit-gangal) [reply go/vt/vtgate/vcursor_impl.go:568]: this needs to go away

File: go/vt/vtgate/safe_session.go [hunk 6]
```diff
@@ -282,7 +287,7 @@ func (session *SafeSession) InTransaction() bool {
 }
 
 // Find returns the transactionId and tabletAlias, if any, for a session
-func (session *SafeSession) Find(keyspace, shard string, tabletType topodatapb.TabletType) (transactionID int64, reservedID int64, alias *topodatapb.TabletAlias) {
+func (session *SafeSession) Find(keyspace, shard string, tabletType topodatapb.TabletType) (int64, int64, *topodatapb.TabletAlias, error) {
```
Comment (systay) [inline go/vt/vtgate/safe_session.go:290]: I didn't find it intuitive that a method with the name `Find` actually changes the state of the shard sessions. Maybe we should rename it?

**Review by systay — APPROVED** [review]

**Pull request merged**